### PR TITLE
feat(auth): add Twitter login for Android, iOS and Web

### PR DESCRIPTION
**Pull request by Lyokone**
## Description

The recommended way of sign in with Twitter was unsafe (since it was exposing secrets to the front)

## Related Issues

*Replace this paragraph with a list of issues related to this PR from the [issue database](https://github.com/flutter/flutter/issues). Indicate, which of these issues are resolved or fixed by this PR. Note that you'll have to prefix the issue numbers with flutter/flutter#.*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [ ] I read and followed the [Flutter Style Guide].
- [ ] I signed the [CLA].
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [X] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/firebase/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: docs/auth/federated-auth.md [hunk 1]
```diff
@@ -223,80 +223,48 @@ import 'package:firebase_auth/firebase_auth.dart';
...
-
-  Create a Twitter provider and provide the credential to the `signInWithPopup` method. This will trigger a new
-  window to appear prompting the user to sign-in to your Twitter application:
-
-  ```dart
-  Future<UserCredential> signInWithTwitter() async {
-    // Create a new provider
-    TwitterAuthProvider twitterProvider = TwitterAuthProvider();
+  Before you begin [configure Sign In with Apple](/docs/auth/web/apple#configure-sign-in-with-apple)
+  and [enable Apple as a sign-in provider](/docs/auth/web/apple#enable-apple-as-a-sign-in-provider).
```
Comment (russellwheatley) [inline docs/auth/federated-auth.md:253]: Is this correct?

**Review by russellwheatley — APPROVED** [review]

File: docs/auth/federated-auth.md (hunk 1 above)
Comment (Lyokone) [reply docs/auth/federated-auth.md]: Good catch 😁 Nothing to do for Web so I forgot to remove the copy paste

File: docs/auth/federated-auth.md [hunk 2]
```diff
@@ -223,80 +223,47 @@ import 'package:firebase_auth/firebase_auth.dart';
...
+    await FirebaseAuth.instance.signInWithAuthProvider(microsoftProvider);
   }
 }
 ```
 
 ## Twitter
 
 Ensure the "Twitter" sign-in provider is enabled on the [Firebase Console](https://console.firebase.google.com/project/_/authentication/providers)
-with an API Key and API Secret set.
+with an API Key and API Secret set. Also make sure your Firebase OAuth redirect URI (e.g. my-app-12345.firebaseapp.com/__/auth/handler)
```
Comment (russellwheatley) [inline docs/auth/federated-auth.md:236]:
```suggestion
with an API Key and API Secret set. Ensure your Firebase OAuth redirect URI (e.g. my-app-12345.firebaseapp.com/__/auth/handler) 
```

File: docs/auth/federated-auth.md [hunk 3]
```diff
@@ -223,80 +223,47 @@ import 'package:firebase_auth/firebase_auth.dart';
...
-    final twitterLogin = new TwitterLogin(
-      apiKey: '<your consumer key>',
-      apiSecretKey:' <your consumer secret>',
-      redirectURI: '<your_scheme>://'
-    );
+* {iOS+}
 
-    // Trigger the sign-in flow
-    final authResult = await twitterLogin.login();
+  You need to configure custom URL scheme as [described in iOS guide step 1](https://firebase.google.com/docs/auth/ios/twitter-login#handle_the_sign-in_flow_with_the_firebase_sdk).
```
Comment (russellwheatley) [inline docs/auth/federated-auth.md:243]:
```suggestion
  You need to configure your custom URL scheme as [described in iOS guide step 1](https://firebase.google.com/docs/auth/ios/twitter-login#handle_the_sign-in_flow_with_the_firebase_sdk).
```

File: packages/firebase_auth/firebase_auth_platform_interface/lib/src/method_channel/utils/convert_auth_provider.dart [hunk 4]
```diff
@@ -24,5 +24,11 @@ AuthProvider convertToOAuthProvider(AuthProvider authProvider) {
     return oAuthProvider;
   }
 
+  if (authProvider is TwitterAuthProvider) {
+    final oAuthProvider = OAuthProvider(authProvider.providerId);
```
Comment (russellwheatley) [inline packages/firebase_auth/firebase_auth_platform_interface/lib/src/method_channel/utils/convert_auth_provider.dart:28]: do we have to set scopes here? `oAuthProvider.setScopes(authProvider.scopes);`

**Review by russellwheatley — CHANGES_REQUESTED** [review]

File: packages/firebase_auth/firebase_auth_platform_interface/lib/src/method_channel/utils/convert_auth_provider.dart (hunk 4 above)
Comment (Lyokone) [reply packages/firebase_auth/firebase_auth_platform_interface/lib/src/method_channel/utils/convert_auth_provider.dart:28]: Twitter does not offer scopes, so it's not available on the TwitterAuthProvider

**Review by russellwheatley — APPROVED** [review]

**Review by Salakar — APPROVED** [review]

**Pull request merged**